paasta_tools/cli/cmds/mesh_status.py

```diff
@@ -0,0 +1,172 @@
+        return (
+            exc.status,
+            [PaastaColors.red(exc.body if exc.status == 405 else exc.reason)],
+        )
+    except (client.connection_error, client.timeout_error) as exc:
+        return (
+            1,
+            [PaastaColors.red(f"Could not connect to API: {exc.__class__.__name__}")],
+        )
+    except Exception as e:
+        output = [PaastaColors.red(f"Exception when talking to the API:")]
+        output.extend(str(e).split("\n"))
+        return 1, output
+
+    output = []
+    if mesh_status.smartstack is not None:
+        smartstack_status_human = get_smartstack_status_human(
+            mesh_status.smartstack.registration,
+            mesh_status.smartstack.expected_backends_per_location,
+            mesh_status.smartstack.locations,
+        )
+        output.extend(smartstack_status_human)
+    if mesh_status.envoy is not None:
+        envoy_status_human = get_envoy_status_human(
+            mesh_status.envoy.registration,
+            mesh_status.envoy.expected_backends_per_location,
+            mesh_status.envoy.locations,
+        )
+        output.extend(envoy_status_human)
+
+    return 0, output
+
+
+def report_mesh_status_for_cluster(
+    cluster: str,
+    service: str,
+    instances: List[str],
+    system_paasta_config: SystemPaastaConfig,
+    verbose: int = 0,
+) -> Tuple[int, Sequence[str]]:
+    output = [f"service: {service}", f"cluster: {cluster}"]
+    return_codes = []
+
+    for instance in instances:
+        return_code, instance_output = paasta_mesh_status_on_api_endpoint(
+            cluster=cluster,
+            service=service,
+            instance=instance,
+            verbose=verbose,
+            system_paasta_config=system_paasta_config,
+        )
+
+        return_codes.append(return_code)
+        output.append(f"  instance: {PaastaColors.cyan(instance)}")
+        output.extend(["    " + line for line in instance_output])
+
+    return 1 if any(return_codes) else 0, output
+
+
+def paasta_mesh_status(args) -> int:
+    system_paasta_config = load_system_paasta_config()
+
+    return_codes = [0]
+    tasks = []
+    clusters_services_instances = apply_args_filters(args)
+    for cluster, service_instances in clusters_services_instances.items():
+        for service, instances in service_instances.items():
+            tasks.append(
+                (
+                    report_mesh_status_for_cluster,
+                    dict(
+                        cluster=cluster,
+                        service=service,
+                        instances=list(instances.keys()),
+                        system_paasta_config=system_paasta_config,
+                        verbose=args.verbose,
+                    ),
+                )
+            )
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
+        tasks = [executor.submit(t[0], **t[1]) for t in tasks]  # type: ignore
```

Comment:
```suggestion
        tasks = [executor.submit(func, **params) for func, params in tasks]  # type: ignore
```
might be a little more self-documenting, but feel free to ignore